asst2/CODE/memory_finder.c

```diff
@@ -0,0 +1,438 @@
+	int first_page_index;
+	int last_page_index;
+	int free_page;
+	int num_free;
+	find_contig_space(curr_id, &free_page, &num_free, &first_page_index, &last_page_index);
+	size_t segments_alloc = size / SEGMENTSIZE + 1;
+	int pages_alloc = (segments_alloc/num_segments+1);
+
+	int alloc_page_index; //page index according the process where an allocation could start
+	int alloc_seg_index = -1; //segment index within alloc_page_index page where allocation could start
+	int contig_free_space = 0; //from start index, to current index how much contiguous free space available
+	int process_index = first_page_index;
+	uint8_t ovf_len = 0;
+	//traverse memory to find the most current current contigious free block that can fit the size requested
+	while (process_index <= last_page_index) {
+		int actual_index = ht_get(ht_space, curr_id, (ht_key)process_index);
+		pagedata* pdata = (pagedata*)myblock + actual_index;
+		metadata* start = (metadata*) (mem_space + actual_index*page_size);
+		int seg_index = ovf_len;
+		while ( seg_index < num_segments) {
+			metadata* curr = start + seg_index*SEGMENTSIZE;
+			uint8_t curr_seg = dm_block_size(curr);
+			if (dm_is_last_segment(curr)) {
+				if (pg_is_overflow(pdata)) ovf_len=curr_seg;
+				else ovf_len = 0;
+				curr_seg = num_segments - seg_index;
+				if (last_page_index == process_index) {
+	//				printf("last page last seg %d\n", seg_index);
+					alloc_seg_index = seg_index;
+				}
+			}
+
+			if (!dm_block_occupied(curr)){
+				if (contig_free_space == 0) {
+					alloc_page_index = process_index;
+					alloc_seg_index = seg_index;
+				}
+				contig_free_space+=curr_seg;
+				if (contig_free_space >= segments_alloc){
+					//current segment can service the request
+					if (curr_seg >= segments_alloc) {
+						dm_allocate_block(curr, segments_alloc);
+						swap(actual_index, process_index);
+						metadata* swapped_mdata = (metadata*) (mem_space+process_index*page_size) + seg_index*SEGMENTSIZE;
+//						printf("segment mallc\n");
+						return (void*) (swapped_mdata + 1);
+					}
+					//multiple page allocation required --> outsource
+					//free_page = -1 to convey that this is
+					//memory allocation within existing spac
+//					printf("contig page alloc\n");
+					free_page = -1;
+					break;
+				}
+			}
+			else {
+				contig_free_space = 0;
+				alloc_page_index = -1;
+//				alloc_seg_index = -1;
+			}
+			seg_index+=curr_seg;
+		}
+
+		if (contig_free_space >= segments_alloc) break;
+		else if (ovf_len) process_index+=(pdata->length-1);
+		else { process_index++; }
+	}
+
+	//if we need to allocate new pages, make sure there's enough space
+	if ( free_page != -1) {
+		if (pages_alloc > num_free) return NULL;
+	//	printf("free page new page\n");
+		alloc_page_index = last_page_index;
+	}
+	return page_allocate(size, curr_id, free_page, alloc_page_index, alloc_seg_index);
+}
+
+/**
+ * Releases pages from process' contiguous space if they are free and at front or end of memory space
```

Comment:
I'm still a little confused what this does

Reply:
It releases pages owned by a process that are compleletly free, ie 64 free segments, back to memory space that another process can malloc, in other words make these pages `NO_OCC`, not occupied. 

But we can't release pages if they are in between the process' view of memory space, the pages have to be at the start or at the end of the process' view of memory space. Let's say a process thinks its memory space spans from index 2 to 10:
- indices 3 to 4 have 64 free segments, we cannot release these pages
- inidices 2 to 5 are completely free, release these pages
- indices 7 to 10, we can release these.

Reply:
why can't we release pages in the middle?

Reply:
if we release from the middle the memory will not be contiguous anymore, it'll be broken

Reply:
and this is a problem because we need to be able to traverse the list to verify pointers for `free()`?

Reply:
i think the main concern is that a process' view of memory needs to be contiguous. and yea everything else i did would not work if the process' view of memory is not contiguous

Reply:
but why not? if there are pages that are not occupied in the middle, can't we just add their length to a free block length and `free` them?